### PR TITLE
Added initial test runner implementation

### PR DESCRIPTION
**Pull request by Bajger**
Added Dockerfile and bin scripts

### CODE REVIEW

**Comment by Bajger** [comment]
@ErikSchierboom: Thank you for reviewing this! Relates to changes in: https://github.com/exercism/pharo-smalltalk/pull/563

**Comment by ErikSchierboom** [comment]
@Bajger Great that you're working on this! What I'd really like to see are some "golden tests". These are tests that contain pre-defined inputs that match the real world situation and those are then combined with their expected output. This makes it very easy for CI to check if the test runner works as intended, and also makes reviewing a lot easier.

You can find example of this in the [generic-test-runner repo](github.com/exercism/generic-test-runner):
- The tests: https://github.com/exercism/generic-test-runner/tree/main/tests
- Script to run the tests: https://github.com/exercism/generic-test-runner/blob/main/bin/run-tests.sh
- Script to run the tests in docker: https://github.com/exercism/generic-test-runner/blob/main/bin/run-tests-in-docker.sh
- CI workflow to run the tests in Docker: https://github.com/exercism/generic-test-runner/blob/main/.github/workflows/ci.yml

**Comment by Bajger** [comment]
> Script to run the tests in docker: https://github.com/exercism/generic-test-runner/blob/main/bin/run-tests-in-docker.sh  

@ErikSchierboom Hi Erik! How do you see using Dockerfile build arguments? Rationale behind: Pharo image for Exercism is constructed from something called "project baseline" The whole project does include all code, but for particular usages we use just sub-parts of Pharo track project using groups. Some examples:
- For test runner functionality without any tests we'd have Dockerfile line:
`RUN pharo metacello install github://exercism/pharo-smalltalk:main/releases/lates BaselineOfExercism --groups=TestRunner`
- But for test runner functionality with all golden test example exercises loaded in (objects in Pharo image) we'd use (differs on last argument): 
`RUN pharo metacello install github://exercism/pharo-smalltalk:main/releases/lates BaselineOfExercism --groups=TestRunner-Tests` - which embraces test runner packages with additional examples in. 

Question: Does Exercism project allows Dockerfile build argument? 
e.g.: `docker build --build-arg groupname=testRunner-tests --rm --no-cache -t pharo-test-runner . `
Thanks for reply!

File: examples/example-all-fail/expected_results.json
Comment (ErikSchierboom) [inline examples/example-all-fail/expected_results.json]:
```suggestion
  "version" : "2",
```

This comment also applies to the other files.

File: examples/example-all-fail/expected_results.json
Comment (ErikSchierboom) [inline examples/example-all-fail/expected_results.json]: Quoting [the docs](https://exercism.org/docs/building/tooling/test-runners/interface):

> key: message, type: string, presence: required if status = error, or when status = fail and version = 1

In this particular case, the tests all ran, only they failed. This means that there should not be be top-level message.

This comment also applies to the other files.

File: examples/example-all-fail/expected_results.json
Comment (ErikSchierboom) [inline examples/example-all-fail/expected_results.json]: The `output` key should only be used for user output. This looks like it wasn't user output. If it isn't and it is helpful to show to the user, it could be part of the `message` field.

See [the docs](https://exercism.org/docs/building/tooling/test-runners/interface):

> The per-test output key should be used to store and output anything that a user deliberately outputs for a test.

This comment also applies to the other files.

File: examples/example-all-pass/expected_results.json
Comment (ErikSchierboom) [inline examples/example-all-pass/expected_results.json]: This key should not be present for the `pass` state.

**Review by ErikSchierboom — COMMENTED** [review]
Excellent work! I left some comments

File: examples/example-all-pass/expected_results.json
Comment (Bajger) [reply examples/example-all-pass/expected_results.json]: @ErikSchierboom: Just to confirm: Top-level "message" key should be just for error status present?

File: examples/example-all-fail/expected_results.json
Comment (Bajger) [reply examples/example-all-fail/expected_results.json]: I'm not really sure I understand purpose of output. So it is used for something, when test method prints out something to user? Any examples? I've included a stack-trace (as a debugging info) of runned test there, if method fails or raises error/exception.

File: examples/example-all-pass/expected_results.json
Comment (Bajger) [reply examples/example-all-pass/expected_results.json]: E.g.: If I have 3 tests: one passes, one fails, one raises error (e.g. exception). What should be top-level error message (if ever)?

Comment (ErikSchierboom) [reply examples/example-all-pass/expected_results.json]: None. So the best way to think about this is as follows:

- If no tests could be executed, for example due to a syntax error or something like that, then the top-level `status` field is `error` and and the `message` should be set and describe the compiler/runtime error
- If any tests could be executed, the top level `status` field will be `pass` or `fail` and no `message` property set

Speaking of errors, if a test fails because the code threw an error/exception, that is usually still a `fail` status for the test. Applying the above logic here, the `error` status should only be used if tests could not be executed. Throwing an error is usually still reported as a failure by the test framework, so that would count as a `fail` test.

File: examples/example-all-fail/expected_results.json
Comment (ErikSchierboom) [reply examples/example-all-fail/expected_results.json]: The purpose of output is to allow students to do print-like "debugging". It's sole purpose is for the student to output things that are then shown in the online editor. Note that most test runners don't support this, as capturing only what a student's code outputs is oftentimes tricky.

> I've included a stack-trace (as a debugging info) of runned test there, if method fails or raises error/exception.

This is great, but should just be part of the `message` field's contents.

**Comment by ErikSchierboom** [comment]
> @ErikSchierboom Hi Erik! How do you see using Dockerfile build arguments? 

We do support this. For your purposes, what I'd do is to set the `ARG` value within the dockerfile to a default value, the one which should be used when deploying the test runner. That way, CI will just work and you can tweak things locally.

File: examples/example-all-pass/expected_results.json
Comment (Bajger) [reply examples/example-all-pass/expected_results.json]: @ErikSchierboom Well, SUnit is quite strict about this distinction, library consider failures only if assertion fails. Errors are considered all raised errors and exceptions during runtime (e.g. division by 0), so that assertion could not be successfully evaluated. I don't think I can influence this (it would mean some tailored modification of SUnit, which isn't worth of it I think).

Comment (Bajger) [reply examples/example-all-pass/expected_results.json]: Anyway, I will modify top-level error logic, so that only syntax errors during code compilation would put top-level error output (that prevents execution of tests).

File: examples/example-all-fail/expected_results.json
Comment (Bajger) [reply examples/example-all-fail/expected_results.json]: Ok, I will move stack-trace info (along with exception message text) into "message" field. I'd probably ignore user-defined output, since usually Smalltalk image doesn't use stdout to print output (it uses its own stream called Transcript). There could be a way that Transcript output would be re-directed and captured to "output" parameter, but it would be hard to distinguish things that are belonging just to executed test (There is no way to limit unrelated code to use Transcript as well).

File: examples/example-all-pass/expected_results.json
Comment (ErikSchierboom) [reply examples/example-all-pass/expected_results.json]: > I don't think I can influence this (it would mean some tailored modification of SUnit, which isn't worth of it I think).

Don't bother, it's fine

File: examples/example-all-fail/expected_results.json
Comment (ErikSchierboom) [reply examples/example-all-fail/expected_results.json]: If it's hard, don't bother with it

**Review by ErikSchierboom — APPROVED** [review]
LGTM!

**Comment by ErikSchierboom** [comment]
I'll fix CI tomorrow (today is my day off)

**Comment by ErikSchierboom** [comment]
@Bajger I've made some progress fixing CI. @iHiD will be doing some configuration work and then we should be close!

**Comment by ErikSchierboom** [comment]
@Bajger CI is failing. Can you see why?

**Comment by Bajger** [comment]
@ErikSchierboom Hi Erik, this is caused by the fact that Pharo image during startup is trying to store new computer UUID on local filesystem. Since there is --read-only argument on docker run command, it causes a trouble. I'll try to find (on our Pharo community discord), if there is a way to really startup image in "read only mode" so process doesn't write any files. I guess `docker run --read-only` is mandatory requirement, right?

**Comment by ErikSchierboom** [comment]
As a matter of fact it isn't yet. Feel free to remove that line, but please add a comment with a to do to investigate if it can be enabled later ob

**Comment by Bajger** [comment]
Hi Erik! I was able to resolve issue with read-only image. Test runner executed for given exercise, but one last problem remaining: Pharo cannot write resulting json to that directory e.g.: 
`Error: '/opt/test-runner/tests/example-all-fail/results.json'` 
Probably some problem with permissions? 
I ran run-tests-in-docker.sh locally (on my Mac) and it worked. So I don't know, if this isn't problem different host environment setup.

**Comment by ErikSchierboom** [comment]
> Hi Erik! I was able to resolve issue with read-only image. Test runner executed for given exercise, but one last problem remaining: Pharo cannot write resulting json to that directory e.g.:
Error: '/opt/test-runner/tests/example-all-fail/results.json'
Probably some problem with permissions?
I ran run-tests-in-docker.sh locally (on my Mac) and it worked. So I don't know, if this isn't problem different host environment setup.

Are you writing `results.json` to the output dir as passed into the `run.sh` file as the third argument?

**Comment by Bajger** [comment]
> Are you writing results.json to the output dir as passed into the run.sh file as the third argument?  

Yes, if you look at run-tests.sh there is:
```
test_dir_name=$(basename "${test_dir}")
        test_dir_path=$(realpath "${test_dir}")
        results_file_path="${test_dir_path}/results.json"
        expected_results_file_path="${test_dir_path}/expected_results.json"

        bin/run.sh "${test_dir_name}" "${test_dir_path}" "${test_dir_path}"
```
and log from Pharo image says following, which corresponds to above expression of run.sh:
```
Loading sources for exercise: 'example-all-fail' from input directory: '/opt/test-runner/tests/example-all-fail', result to output directory: '/opt/test-runner/tests/example-all-fail'.
Error: '/opt/test-runner/tests/example-all-fail/results.json'
```
It seems that mounted directory from run-tests-in-docker.sh from host file-system is not writeable:
line: `--mount type=bind,src="$PWD/tests",dst=/opt/test-runner/tests \`

**Comment by ErikSchierboom** [comment]
> It seems that mounted directory from run-tests-in-docker.sh from host file-system is not writeable

Weird, as we're doing the same thing with other test runners. I'll look into it.

**Comment by Bajger** [comment]
I can try to create / touch some file using shell command (part of run-tests.sh) to proof that it is not writeable. Let me know!

**Comment by ErikSchierboom** [comment]
@Bajger That would be really useful yes!

**Comment by Bajger** [comment]
@ErikSchierboom Here it is: Lines in commit above resulted in: 
```
/opt/test-runner/bin/run-tests.sh: 24: cannot create /opt/test-runner/tests/example-all-fail/myfile.txt: Permission denied
cat: /opt/test-runner/tests/example-all-fail/myfile.txt: No such file or directory
Printing content of: /opt/test-runner/tests/example-all-fail/myfile.txt
```

**Comment by Bajger** [comment]
Not sure if different owner of test directory is something that might prevent to write file? From build log:
```
The owner of the /opt/test-runner/tests/example-all-raise-error directory is UNKNOWN.
The current user is pharo
```

**Comment by ErikSchierboom** [comment]
It must be due to this:

```
# remove unncecessary changes file of Pharo
RUN rm /opt/test-runner/Pharo.changes

# set owner to pharo user
RUN chown -R pharo /opt/test-runner 

# set write permission for pharo user
RUN chmod -R u+w /opt/test-runner  

USER pharo
```

The `pharo` user will have the wrong permissions, as the chmodding done above doesn't apply to the tests directory that is later "plugged in"

**Comment by Bajger** [comment]
@ErikSchierboom So I should remove chown, chmod lines, right?

**Comment by ErikSchierboom** [comment]
@Bajger Well, they don't really do anything I think. Not that removing those solves the issue

**Comment by ErikSchierboom** [comment]
I've pushed a commit to saves the test results in a /tmp dir, which the `pharo` user _can_ write to: https://github.com/exercism/pharo-smalltalk-test-runner/pull/16/commits/5ee60a7d5dfa6fc937a63a4daafb2b33708670f0

**Comment by Bajger** [comment]
Ah, sorry, I was a bit interfering with changes. I thought to specify UID for user and then use uid for mount command in docker run.

**Comment by ErikSchierboom** [comment]
I think progress was made! Still fails, but better messages

**Comment by Bajger** [comment]
Right, we are getting closer! I'm a bit confused with error message, since example exercises should be part of Pharo image.
You can see: docker build should use line: `#8 [loader 2/3] RUN pharo metacello install github://bajger/pharo-smalltalk:453-Build-Test-Runner/dev/src BaselineOfExercism --groups=testRunnerTests`). But there is prior build stage (so probably wrong docker image is used, intead of image created as part of run-tests-in-docker.sh?): 
image

TLDR: I think wrong docker image is used in run-tests-in-docker.sh (it probably uses created image in stage before, which uses testRunner baseline and not testRunnerTests baseline)

**Comment by ErikSchierboom** [comment]
> TLDR: I think wrong docker image is used in run-tests-in-docker.sh (it probably uses created image in stage before, which uses testRunner baseline and not testRunnerTests baseline)

Is this something you can fix?

**Comment by Bajger** [comment]
I'm not the Docker or Github action expert, but it really looks that wrong docker image is used. CI steps creates one into pushes into cache (production image - without build args): 
https://github.com/exercism/pharo-smalltalk-test-runner/actions/runs/4194593506/jobs/7272960773#step:4:117

and then another (with build args - it creates example test image) is built in run-tests-in-docker.sh here:
https://github.com/exercism/pharo-smalltalk-test-runner/actions/runs/4194593506/jobs/7272960773#step:5:5

So I don't know, how to really convince script to use/run docker image built with args (and not that one from cache).

**Comment by Bajger** [comment]
@ErikSchierboom Hooray! 🎉 It seems that CI is green (the --load argument did work). Please check the GH action output, if docker image is properly deployed.

**Comment by Bajger** [comment]
I will change correct repository reference in: https://github.com/Bajger/pharo-smalltalk-test-runner/blob/09ad0cc250165cbae483dd30b459000ca9f10936/Dockerfile#L3
once https://github.com/exercism/pharo-smalltalk/pull/563 is integrated.

**Comment by ErikSchierboom** [comment]
Awesome! To enable the test runner, you'll need to update the track's config.json file.
See https://github.com/exercism/zig/commit/72918d7873e0828646fb906da15a9edd61a16b01 for an example on how to do that.

**Comment by Bajger** [comment]
Thanks a lot for your patience and effort! I will enable test runner, once related PR is integrated (waiting for review from Glenn): https://github.com/exercism/pharo-smalltalk/pull/563

**Pull request merged**
